### PR TITLE
Add support for test target creation in the pods project generator

### PR DESCRIPTION
**Pull request by dnkoutso**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -25,7 +25,7 @@ gem 'json', :git => 'https://github.com/segiddins/json.git', :branch => 'seg-1.7
```
Comment (dnkoutso) [inline Gemfile]: will be reverted.

File: Gemfile.lock [hunk 2]
```diff
@@ -6,16 +6,6 @@ GIT
     claide (1.0.1)
```
Comment (dnkoutso) [inline Gemfile.lock]: will be reverted.

File: lib/cocoapods/generator/xcconfig/aggregate_xcconfig.rb [hunk 3]
```diff
@@ -176,40 +177,6 @@ def generate_settings_to_import_pod_targets
           end
         end
 
-        # Add custom build settings and required build settings to link to
```
Comment (dnkoutso) [inline lib/cocoapods/generator/xcconfig/aggregate_xcconfig.rb]: Those moved to `xcconfig_helper` no changes here and specs pass.

File: lib/cocoapods/generator/xcconfig/pod_xcconfig.rb [hunk 4]
```diff
@@ -66,6 +70,11 @@ def generate
           end
           XCConfigHelper.add_target_specific_settings(target, @xcconfig)
           @xcconfig.merge! XCConfigHelper.settings_for_dependent_targets(target, target.recursive_dependent_targets)
+          if @test_xcconfig
```
Comment (dnkoutso) [inline lib/cocoapods/generator/xcconfig/pod_xcconfig.rb]: For test `xcconfigs` we ensure that linker flags are added for their dependencies.

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 5]
```diff
@@ -45,13 +45,15 @@ def self.quote(strings, prefix = nil)
...
+        #         whether to include `-ObjC` in the other linker flags
         #
-        def self.default_ld_flags(target, includes_static_libraries = false)
+        # @return [String] the default linker flags. `-ObjC` is optionally included depending
+        #         on the target while `-fobjc-arc` is included only if requested in the Podfile.
+        #
+        def self.default_ld_flags(target, include_objc_flag = false)
           ld_flags = ''
-          ld_flags << '-ObjC' if includes_static_libraries
+          ld_flags << '-ObjC' if include_objc_flag
```
Comment (dnkoutso) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: just rename of the flag here no real changes.

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 6]
```diff
@@ -296,6 +298,55 @@ def self.settings_for_dependent_targets(target, dependent_targets)
           build_settings
         end
 
+        # Add custom build settings and required build settings to link to
```
Comment (dnkoutso) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: copied pasted from `aggregate_xcconfig.rb`, almost zero changes except passing parameters.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 7]
```diff
@@ -86,6 +86,11 @@ def share_development_pod_schemes
           development_pod_targets.select(&:should_build?).each do |pod_target|
             next unless share_scheme_for_development_pod?(pod_target.pod_name)
             Xcodeproj::XCScheme.share_scheme(project.path, pod_target.label)
+            if pod_target.contains_test_specifications?
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: Nicely shares the `xcscheme` for tests :D

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 8]
```diff
@@ -203,21 +207,32 @@ def set_target_dependencies
...
-                  pod_target.native_target.frameworks_build_phase.add_file_reference(product_ref, true)
-                end
+              pod_target.test_native_targets.each do |test_native_target|
+                add_dependent_targets_to_native_target(pod_target.test_dependent_targets, test_native_target, false, pod_target.requires_frameworks?)
               end
             end
           end
         end
 
+        # TODO
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: will take care of this

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 9]
```diff
@@ -18,20 +18,23 @@ def install!
...
               create_support_files_dir
               add_resources_bundle_targets
               add_files_to_build_phases
               create_xcconfig_file
+              create_test_xcconfig_files if target.contains_test_specifications?
               if target.requires_frameworks?
                 create_info_plist_file
                 create_module_map
                 create_umbrella_header do |generator|
+                  file_accessors = target.file_accessors.reject { |f| f.spec.test_specification? }
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: This prevents adding test specification headers as part of the generated umbrella header

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 10]
```diff
@@ -162,6 +166,68 @@ def add_files_to_build_phases
...
+          def product_type_for_test_type(test_type)
+            case test_type
+            when :unit
+              :unit_test_bundle
+            else
+              raise Informative, "Unknown test type passed #{test_type}."
+            end
+          end
+
+          # TODO: doc
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: will fix

File: lib/cocoapods/target/pod_target.rb [hunk 11]
```diff
@@ -234,7 +267,7 @@ def include_in_build_config?(target_definition, configuration_name)
 
       if whitelists.empty?
         @build_config_cache[key] = true
-        return true
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: just warning fix

File: lib/cocoapods/target/pod_target.rb [hunk 12]
```diff
@@ -258,7 +291,7 @@ def inhibit_warnings?
 
       if whitelists.empty?
         @inhibit_warnings = false
-        return false
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: just warning fix

File: lib/cocoapods/target/pod_target.rb [hunk 13]
```diff
@@ -268,7 +301,7 @@ def inhibit_warnings?
           'settings to inhibit warnings. CocoaPods does not currently ' \
           'support different settings and will fall back to your preference ' \
           'set in the root target definition.'
-        return podfile.root_target_definitions.first.inhibits_warnings_for_pod?(root_spec.name)
+        podfile.root_target_definitions.first.inhibits_warnings_for_pod?(root_spec.name)
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: just warning fix

File: spec/fixtures/coconut-lib/CoconutLib.podspec [hunk 14]
```diff
@@ -0,0 +1,20 @@
+Pod::Spec.new do |s|
```
Comment (dnkoutso) [inline spec/fixtures/coconut-lib/CoconutLib.podspec]: New library that helps with `test_spec` support

**Comment by dnkoutso** [comment]
@benasher44, @DanToml, @orta (and maybe @segiddins?) would love for you to review this PR at your time.

I need to add a few more specs but overall its 95% complete.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 15]
```diff
@@ -203,16 +208,10 @@ def set_target_dependencies
               aggregate_target.native_target.add_dependency(pod_target.native_target)
               configure_app_extension_api_only_for_target(pod_target) if is_app_extension
 
-              pod_target.dependent_targets.each do |pod_dependency_target|
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: method extracted. No logical changes. Helps support wiring up test dependencies to the native targets

**Comment by orta** [comment]
@dnkoutso any chance you can upload a zip of an example project with this?

**Comment by dnkoutso** [comment]
@orta will upload one today

**Comment by dnkoutso** [comment]
@orta there you go.

Please keep in mind I am still using a `Podfile` to consume the test spec since the validator work is not done.

You will also need to ensure you setup cocoapods to point to my repo and branch so the integration works.

It's pretty awesome to see it in action :)

[SampleProject.zip](https://github.com/CocoaPods/CocoaPods/files/994071/SampleProject.zip)

**Comment by dnkoutso** [comment]
Added a few more specs!

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! 
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* Add support for test target creation in the pods project generator 
  [dnkoutso](https://github.com/dnkoutso)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by dnkoutso** [comment]
Should go green!

**Comment by dnkoutso** [comment]
Green!

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 16]
```diff
@@ -400,7 +489,7 @@ def header_mappings_dir
                                    end
           end
 
-          def add_header(build_file, public_headers, private_headers)
+          def add_header(build_file, public_headers, private_headers, native_target)
```
Comment (DanToml) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: this added param is not used

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 9 above)
Comment (DanToml) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: you may want to gate the reject on contains_test_specification - might be expensive to do this for larger projects that do not have test specs

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 16 above)
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: It is, the method uses `native_target` which is a property but now the parameter takes over so no real method changes were required.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 9 above)
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: 👍 nice thanks for the tip will do

Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: @DanToml done, also added a test case to ensure test header files never make it into umbrella header :)

**Comment by dnkoutso** [comment]
All set.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 16 above)
Comment (DanToml) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: ahhh - lol shadowing

**Comment by orta** [comment]
I wonder if we can do something about making `cmd + u` work for the project. Right now, tests only work inside the actual test target.

I wonder if we can make it work both for the library:

So for this setup: 
screen shot 2017-05-15 at 10 37 06

Ideally `cmd + u` on Pods-SampleProject, and SampleLibrary would also work. It would run any tests for any pod-targets that are linked to the "cocoapods target"

**Comment by DanToml** [comment]
@orta Yeah it would be pretty nice if we added the testable to the main scheme

**Comment by dnkoutso** [comment]
Yes! So CMD+U works today on the test scheme but *only if* you decide to have `share_development_pod_schemes` enabled.

Second, Xcodeproj has a method called `recreate_user_schemes` so I tried to edit the xcconfigs but there we being "nuked" again and re-written by this method.

I think we should expose a block in the Xcodeproj method and allow us to perform various customization on the schemes before we save them, so we don't have to do a post-process step and save them twice which can be slow for multiple files.

Is this a blocker in this PR?

**Comment by orta** [comment]
IMO, ideally that gets looked at before the RC releases, but I don't consider it a blocker on this PR which is "get it in" vs "make it perfect"

**Comment by dnkoutso** [comment]
With your blessing we can get this merged and start a bit the process of polishing it!

I got an upcoming branch with validator changes for running tests during `lint`

**Comment by dnkoutso** [comment]
<3!

**Comment by segiddins** [comment]
🍾 well done, @dnkoutso !

**Pull request merged**